test/common/router/router_test.cc

```diff
@@ -2227,6 +2227,190 @@ TEST_F(RouterTest, DirectResponseWithBody) {
+  ON_CALL(callbacks_, streamInfo()).WillByDefault(ReturnRef(stream_info));
+  EXPECT_FALSE(stream_info.firstUpstreamTxByteSent().has_value());
+  EXPECT_FALSE(stream_info.lastUpstreamTxByteSent().has_value());
+  EXPECT_FALSE(stream_info.firstUpstreamRxByteReceived().has_value());
+  EXPECT_FALSE(stream_info.lastUpstreamRxByteReceived().has_value());
+
+  Http::TestHeaderMapImpl headers{};
+  HttpTestUtility::addDefaultHeaders(headers);
+  router_.decodeHeaders(headers, false);
+
+  test_time_.sleep(std::chrono::milliseconds(32));
+  Buffer::OwnedImpl data;
+  router_.decodeData(data, true);
+
+  Http::HeaderMapPtr response_headers(new Http::TestHeaderMapImpl{{":status", "503"}});
+  response_decoder->decodeHeaders(std::move(response_headers), false);
+  test_time_.sleep(std::chrono::milliseconds(43));
+
+  // Confirm we still have no upstream timing data. It won't be set until after the
+  // stream has ended.
+  EXPECT_FALSE(stream_info.firstUpstreamTxByteSent().has_value());
+  EXPECT_FALSE(stream_info.lastUpstreamTxByteSent().has_value());
+  EXPECT_FALSE(stream_info.firstUpstreamRxByteReceived().has_value());
+  EXPECT_FALSE(stream_info.lastUpstreamRxByteReceived().has_value());
+
+  response_decoder->decodeData(data, true);
+
+  // Now these should be set.
+  EXPECT_TRUE(stream_info.firstUpstreamTxByteSent().has_value());
+  EXPECT_TRUE(stream_info.lastUpstreamTxByteSent().has_value());
+  EXPECT_TRUE(stream_info.firstUpstreamRxByteReceived().has_value());
+  EXPECT_TRUE(stream_info.lastUpstreamRxByteReceived().has_value());
+
+  // Timings should match our sleep() calls.
+  EXPECT_EQ(stream_info.lastUpstreamRxByteReceived().value() -
+                stream_info.firstUpstreamRxByteReceived().value(),
+            std::chrono::milliseconds(43));
+  EXPECT_EQ(stream_info.lastUpstreamTxByteSent().value() -
+                stream_info.firstUpstreamTxByteSent().value(),
+            std::chrono::milliseconds(32));
+}
+
+// Verify that upstream timing information is set into the StreamInfo when a
+// retry occurs (and not before).
+TEST_F(RouterTest, UpstreamTimingRetry) {
+  NiceMock<Http::MockStreamEncoder> encoder;
+  Http::StreamDecoder* response_decoder = nullptr;
+  EXPECT_CALL(cm_.conn_pool_, newStream(_, _))
+      .WillOnce(Invoke([&](Http::StreamDecoder& decoder, Http::ConnectionPool::Callbacks& callbacks)
+                           -> Http::ConnectionPool::Cancellable* {
+        response_decoder = &decoder;
+        callbacks.onPoolReady(encoder, cm_.conn_pool_.host_);
+        return nullptr;
+      }));
+  expectResponseTimerCreate();
+
+  StreamInfo::StreamInfoImpl stream_info(test_time_);
+  ON_CALL(callbacks_, streamInfo()).WillByDefault(ReturnRef(stream_info));
+
+  // Check that upstream timing is updated after the first request.
+  Http::TestHeaderMapImpl headers{{"x-envoy-retry-on", "5xx"}};
+  HttpTestUtility::addDefaultHeaders(headers);
+  router_.decodeHeaders(headers, false);
+
+  router_.retry_state_->expectRetry();
```

Comment:
Where is expectRetry() declared? Is there some file that was not included in the PR? router_test fails to build in a clean client for me.  AFAICT this is the first mention of this method.

There's also a missing include of assert.h in include/envoy/stream_info/stream_info.h but that was easy to remedy. I'm not sure what the solution is here.